src/bin/neovim_lspc.rs

```diff
@@ -0,0 +1,232 @@
+                .next_element()?
+                .ok_or_else(|| de::Error::invalid_length(1, &self))?;
+            let error: Value = seq
+                .next_element()?
+                .ok_or_else(|| de::Error::invalid_length(2, &self))?;
+            let result: Value = seq
+                .next_element()?
+                .ok_or_else(|| de::Error::invalid_length(3, &self))?;
+
+            Ok(RpcResponse {
+                msgid,
+                error,
+                result,
+            })
+        } else if tag == 2 {
+            let method: String = seq
+                .next_element()?
+                .ok_or_else(|| de::Error::invalid_length(1, &self))?;
+            let params: Vec<Value> = seq
+                .next_element()?
+                .ok_or_else(|| de::Error::invalid_length(2, &self))?;
+
+            Ok(RpcNotification { method, params })
+        } else {
+            Err(de::Error::invalid_value(de::Unexpected::Other("invalid tag"), &self))
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for RpcMessage {
+    fn deserialize<D>(deserializer: D) -> Result<RpcMessage, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(RpcVisitor)
+    }
+}
+
+use RpcMessage::*;
+
+struct Client {
+    tx: Sender<RpcMessage>,
+    tx_thread: JoinHandle<()>,
+    rx_thread: JoinHandle<()>,
+    msgid_counter: u64,
+}
+
+impl Client {
+    fn new<R, W>(mut reader: R, mut writer: W) -> (Self, Receiver<RpcMessage>)
+    where
+        R: Read + Send + 'static,
+        W: Write + Send + 'static,
+    {
+        let (tx, io_rx): (Sender<RpcMessage>, Receiver<RpcMessage>) = channel::unbounded();
+        let (io_tx, rx): (Sender<RpcMessage>, Receiver<RpcMessage>) = channel::unbounded();
+        let rx_thread = thread::spawn(move || {
+            for msg in io_rx {
+                // TODO: Handle err! Log is good
+                rmp_serde::encode::write(&mut writer, &msg).unwrap();
+                writer.flush().unwrap();
+                let mut buf = Vec::new();
+                msg.serialize(&mut Serializer::new(&mut buf)).unwrap();
+                eprintln!("Sent: {:?}", buf);
+            }
+        });
+
+        let tx_thread = thread::spawn(move || {
+            let mut value_reader = Deserializer::new(reader);
+            while let Ok(value) = Deserialize::deserialize(&mut value_reader) {
+                eprintln!("Received: {:?}", value);
+                io_tx.send(value).unwrap();
+            }
+        });
+
+        let msgid_counter = 0;
+
+        (Client {
+            tx,
+            tx_thread,
+            rx_thread,
+            msgid_counter,
+        }, rx)
+    }
+
+    fn send_request(&mut self, method: &str, params: Vec<Value>) {
+        self.msgid_counter += 1;
+        let request = RpcRequest {
+            msgid: self.msgid_counter,
+            method: method.into(),
+            params,
+        };
+        self.tx.send(request).unwrap();
```

Comment:
Hi a.
I think it will be better than if we use explicitly message for showing the error. We can easily to find the code error by the error message.
`self.tx.send(request).expect("Send request failed ...")`